katsdpcontroller/test/test_product_config.py

```diff
@@ -1,73 +1,1012 @@
         orig = {"a": {"aa": 1, "ab": 2}, "b": {"ba": {"c": 10}, "bb": [5]}}
         override = {"a": {"aa": [], "ab": None, "ac": 3}, "b": {"bb": [1, 2]}}
         out = product_config.override(orig, override)
         assert_equal({"a": {"aa": [], "ac": 3}, "b": {"ba": {"c": 10}, "bb": [1, 2]}}, out)
 
 
+class TestUrlNEndpoints:
+    """Test :meth:`~katsdpcontroller.product_config._url_n_endpoints`."""
+
+    def test_simple(self) -> None:
+        assert_equal(product_config._url_n_endpoints('spead://239.1.2.3+7:7148'), 8)
+        assert_equal(product_config._url_n_endpoints('spead://239.1.2.3:7148'), 1)
+
+    def test_yarl_url(self) -> None:
+        assert_equal(product_config._url_n_endpoints(yarl.URL('spead://239.1.2.3+7:7148')), 8)
+        assert_equal(product_config._url_n_endpoints(yarl.URL('spead://239.1.2.3:7148')), 1)
+
+    def test_not_spead(self) -> None:
+        with assert_raises_regex(ValueError, 'non-spead URL http://239.1.2.3:7148'):
+            product_config._url_n_endpoints('http://239.1.2.3:7148')
+
+    def test_missing_part(self) -> None:
+        with assert_raises_regex(ValueError, 'URL spead:/path has no host'):
+            product_config._url_n_endpoints('spead:/path')
+        with assert_raises_regex(ValueError, 'URL spead://239.1.2.3 has no port'):
+            product_config._url_n_endpoints('spead://239.1.2.3')
+
+
+class TestNormaliseOutputChannels:
+    """Test :meth:`~katsdpcontroller.product_config.normalise_output_channels`."""
+
+    def test_given(self) -> None:
+        assert_equal(product_config._normalise_output_channels(100, (20, 80)), (20, 80))
+
+    def test_absent(self) -> None:
+        assert_equal(product_config._normalise_output_channels(100, None), (0, 100))
+
+    def test_empty(self) -> None:
+        with assert_raises_regex(ValueError, r'output_channels is empty \(50:50\)'):
+            product_config._normalise_output_channels(100, (50, 50))
+        with assert_raises_regex(ValueError, r'output_channels is empty \(60:50\)'):
+            product_config._normalise_output_channels(100, (60, 50))
+
+    def test_overflow(self) -> None:
+        with assert_raises_regex(ValueError,
+                                 r'output_channels \(0:101\) overflows valid range 0:100'):
+            product_config._normalise_output_channels(100, (0, 101))
+        with assert_raises_regex(ValueError,
+                                 r'output_channels \(-1:1\) overflows valid range 0:100'):
+            product_config._normalise_output_channels(100, (-1, 1))
+
+    def test_align(self) -> None:
+        assert_equal(product_config._normalise_output_channels(1000, (299, 301), 100), (200, 400))
+        assert_equal(product_config._normalise_output_channels(1000, (200, 400), 100), (200, 400))
+
+    def test_misalign(self) -> None:
+        with assert_raises_regex(
+                ValueError,
+                r'n_chans \(789\) is not a multiple of required alignment \(100\)'):
```

Comment:
Just a question: Testing against an exact human readable error string always seems quite brittle to me. Wouldn't it be better to organise the testing so you cover the unrelated ValueError cases in another way, thus giving you confidence that the ValueError here is the intended one. Or better still, for calls that you can't guarantee unambiguous exceptions, rather have subclasses exceptions.

Reply:
I decided to put the full string in mostly because I have a bad habit of returning error strings that contain curly braces because I forgot to call `.format` on them (or make them f-strings, in 3.6+). This ensures not just that I'm covering the right piece of code, but that the error string is actually useful to a human.

Reply:
You still haven't quite sold me on this. In general I would not really expect changing the log message of an exception slightly to result in the unit test failing. Not sure there is a better solution though... Maybe @ludwigschwardt has a view ?

Reply:
The only other option to check exceptions that I can think of is to return some of these numbers as exception arguments, but that might be overkill again.

Reply:
That still wouldn't guarantee that the exception *message* is properly formatted.

Reply:
We could have a messages library which at least provides a single source for these text strings, but by that stage we are multiple coats deep on the bikeshed. I wave the white flag of defeat :)